### PR TITLE
Changed targeting from .NET 4.5 to .NET 4.5.2

### PR DESCRIPTION
**Pull request by michaelnoonan**
I started down this path because I couldn't build on my machine.
And I couldn't find the .NET 4.5 SDK, short of installing VS2017.
All the same OS support all patches of .NET 4.5 including .NET 4.5.2.
I cleaned up all the build/pack warnings.
I fixed the PowerShell E2E test to handle spaces in the path to the .ps1 file.

# Result

```
D:.
|   Octopus.Client.nuspec
|   [Content_Types].xml
|
+---images
|       icon.png
|
+---lib
|   +---net452
|   |       Octopus.Client.dll
|   |       Octopus.Client.xml
|   |
|   \---netstandard2.0
|           Octopus.Client.dll
|           Octopus.Client.xml
|
+---package
|   \---services
|       \---metadata
|           \---core-properties
|                   6b3ce88c0c45481b817b7dbbee2806d7.psmdcp
|
\---_rels
        .rels
```

NUSPEC file
```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2013/05/nuspec.xsd">
  <metadata>
    <id>Octopus.Client</id>
    <version>8.4.1-ci4</version>
    <authors>Octopus Deploy</authors>
    <owners>Octopus Deploy</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <license type="expression">Apache-2.0</license>
    <licenseUrl>https://licenses.nuget.org/Apache-2.0</licenseUrl>
    <icon>images\icon.png</icon>
    <projectUrl>https://github.com/OctopusDeploy/OctopusClients</projectUrl>
    <description>Octopus Deploy is an automated release management tool for modern developers and DevOps teams. 
            This package contains the client library for the HTTP API in Octopus.</description>
    <copyright>Octopus Deploy Pty Ltd</copyright>
    <repository type="git" url="https://github.com/OctopusDeploy/OctopusClients" />
    <dependencies>
      <group targetFramework=".NETFramework4.5.2" />
      <group targetFramework=".NETStandard2.0">
        <dependency id="Microsoft.CSharp" version="4.5.0" exclude="Build,Analyzers" />
        <dependency id="System.ComponentModel.Annotations" version="4.5.0" exclude="Build,Analyzers" />
        <dependency id="Newtonsoft.Json" version="11.0.2" />
        <dependency id="Octodiff" version="1.2.1" />
      </group>
    </dependencies>
    <frameworkAssemblies>
      <frameworkAssembly assemblyName="System.ComponentModel.DataAnnotations" targetFramework=".NETFramework4.5.2" />
      <frameworkAssembly assemblyName="System.Net.Http" targetFramework=".NETFramework4.5.2" />
      <frameworkAssembly assemblyName="System.Numerics" targetFramework=".NETFramework4.5.2" />
    </frameworkAssemblies>
  </metadata>
</package>
```

### CODE REVIEW

File: source/Octopus.Client.E2ETests/NuSpecDependenciesFixture.cs [hunk 1]
```diff
@@ -78,37 +78,37 @@ public void NuSpecFileShouldHaveADependencyGroupForEachTargetFramework()
         [Test]
         public void NuSpecFileShouldHaveANetFrameworkDependencyGroup()
         {
-            var net45DependencyGroup = nuSpecFile.SelectSingleNode(
```
Comment (michaelnoonan) [inline source/Octopus.Client.E2ETests/NuSpecDependenciesFixture.cs:81]: Simple rename

File: source/Octopus.Client.E2ETests/PowerShellExecutionFixture.cs [hunk 2]
```diff
@@ -26,7 +26,7 @@ public void CanBeRunFromPowerShell()
             var extractedDll = ExtractDll(tempFolder);
             var powerShellFile = CreatePowerShellFile(extractedDll, tempFolder);
 
-            var output = SilentProcessRunner.ExecuteCommand("powershell", $" -NoProfile -NoLogo -NonInteractive -WindowStyle Hidden -ExecutionPolicy Unrestricted -File {powerShellFile}");
+            var output = SilentProcessRunner.ExecuteCommand("powershell", $" -NoProfile -NoLogo -NonInteractive -WindowStyle Hidden -ExecutionPolicy Unrestricted -File \"{powerShellFile}\"");
```
Comment (michaelnoonan) [inline source/Octopus.Client.E2ETests/PowerShellExecutionFixture.cs:29]: This fixed a bug where the `powerShellFile` (path) contained spaces.

File: source/Octopus.Client.Tests/Integration/HttpIntegrationTestBase.cs [hunk 3]
```diff
@@ -91,7 +91,9 @@ private static X509Certificate2 GetCert()
             using (var ms = new MemoryStream())
             {
                 s.CopyTo(ms);
+#pragma warning disable PC001
```
Comment (michaelnoonan) [inline source/Octopus.Client.Tests/Integration/HttpIntegrationTestBase.cs:94]: This is a warning we can safely ignore: `Integration\HttpIntegrationTestBase.cs(94,24): warning PC001: X509Certificate2.X509Certificate2(byte[], string) isn't supported on macOS [D:\dev\OctopusClients\source\Octopus.Client.Tests\Octopus.Client.Tests.csproj]`

File: source/Octopus.Client.Tests/Octopus.Client.Tests.csproj [hunk 4]
```diff
@@ -44,7 +44,7 @@
     <PackageReference Include="Autofac" Version="4.6.0" />
     <PackageReference Include="FluentAssertions" Version="4.15.0" />
     <PackageReference Include="Nancy" Version="2.0.0" />
-    <PackageReference Include="NSubstitute" Version="2.0.3" />
+    <PackageReference Include="NSubstitute" Version="4.2.1" />
```
Comment (michaelnoonan) [inline source/Octopus.Client.Tests/Octopus.Client.Tests.csproj:47]: Upgraded NSubstitute to remove build warnings.

File: source/Octopus.Client/Octopus.Client.csproj [hunk 5]
```diff
@@ -5,32 +5,26 @@
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <AssemblyName>Octopus.Client</AssemblyName>
     <GenerateAssemblyTitleAttribute>false</GenerateAssemblyTitleAttribute>
-    <PackageId>Octopus.Client</PackageId>
```
Comment (michaelnoonan) [inline source/Octopus.Client/Octopus.Client.csproj:8]: All the package stuff was duplicated in the .nuspec file, and this was generating warnings.

File: source/Octopus.Client/Octopus.Client.nuspec [hunk 6]
```diff
@@ -6,15 +6,15 @@
         <authors>Octopus Deploy</authors>
         <owners>Octopus Deploy</owners>
         <requireLicenseAcceptance>false</requireLicenseAcceptance>
-        <licenseUrl>https://github.com/OctopusDeploy/OctopusClients/blob/master/LICENSE.txt</licenseUrl>
-        <projectUrl>http://octopusdeploy.com/</projectUrl>
-        <iconUrl>http://i.octopusdeploy.com/resources/Avatar3-32x32.png</iconUrl>
+        <license type="expression">Apache-2.0</license>
```
Comment (michaelnoonan) [inline source/Octopus.Client/Octopus.Client.nuspec:9]: `licenseUrl` is deprecated: https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5125

File: source/Octopus.Client/Octopus.Client.nuspec [hunk 7]
```diff
@@ -6,15 +6,15 @@
...
+        <projectUrl>https://github.com/OctopusDeploy/OctopusClients</projectUrl>
+        <icon>images\icon.png</icon>
```
Comment (michaelnoonan) [inline source/Octopus.Client/Octopus.Client.nuspec:11]: `iconUrl` is deprecated: https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5048

**Comment by michaelnoonan** [comment]
The test failure was flaky. Killing the agent and starting again got the tests running.

File: source/Octopus.Client/Octopus.Client.nuspec [hunk 8]
```diff
@@ -24,14 +24,15 @@
...
-            <frameworkAssembly assemblyName="System.Numerics" targetFramework=".NETFramework4.5" />
+            <frameworkAssembly assemblyName="System.ComponentModel.DataAnnotations" targetFramework=".NETFramework4.5.2" />
+            <frameworkAssembly assemblyName="System.Net.Http" targetFramework=".NETFramework4.5.2" />
+            <frameworkAssembly assemblyName="System.Numerics" targetFramework=".NETFramework4.5.2" />
         </frameworkAssemblies>
     </metadata>
     <files>
-        <file src="bin\release\net45\Octopus.Client.dll" target="lib\net45" />
-        <file src="bin\release\net45\Octopus.Client.xml" target="lib\net45" />
+        <file src="icon.png" target="images\" />
```
Comment (johnsimons) [inline source/Octopus.Client/Octopus.Client.nuspec:33]: Is this needed @michaelnoonan ?

File: source/Octopus.Client/app.config [hunk 9]
```diff
@@ -9,6 +9,6 @@
     </assemblyBinding>
   </runtime>
   <startup>
-    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5"/>
+    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5.2"/>
```
Comment (johnsimons) [inline source/Octopus.Client/app.config:12]: Do we even need this?

File: source/Octopus.Client/Octopus.Client.nuspec (hunk 7 above)
Comment (johnsimons) [inline source/Octopus.Client/Octopus.Client.nuspec:11]: ah I see now, so the icon is now part of the package

File: source/Octopus.Client/Octopus.Client.nuspec (hunk 8 above)
Comment (michaelnoonan) [reply source/Octopus.Client/Octopus.Client.nuspec:33]: Yes. It's related to `iconUrl` being deprecated. Embedding an image into the package is the recommended approach.

File: source/Octopus.Client/Octopus.Client.nuspec (hunk 7 above)
Comment (johnsimons) [reply source/Octopus.Client/Octopus.Client.nuspec:11]: is this the way forward for nugets?

File: source/Octopus.Client/app.config (hunk 9 above)
Comment (michaelnoonan) [reply source/Octopus.Client/app.config:12]: That's a good question. I am unsure to be honest. 👀

**Comment by johnsimons** [comment]
I wonder if we could get rid of the nuspec and autogen it as part of the csproj, less maintenance

**Comment by michaelnoonan** [comment]
I wonder the same. I couldn't see a recommendation from Microsoft either way. I like the explicit nature of the .nuspec file - it's only job in the world is to configure the packaging. The csproj file already has a lot of jobs. In my research I discovered tooling like Rider uses the deprecated properties in the csproj file:
image

**Comment by johnsimons** [comment]
The beauty of doing it as part of the csproj is that there is only one place to maintain dependencies and output artifacts

File: source/Octopus.Client/app.config (hunk 9 above)
Comment (johnsimons) [reply source/Octopus.Client/app.config:12]: We removed all these from other projects, but they are no longer targeting a mix of netfull + netcore

Comment (michaelnoonan) [reply source/Octopus.Client/app.config:12]: I think we can remove the whole thing. The documentation only talks about "applications". This [blog post supports](https://stackoverflow.com/questions/21566528/what-happens-if-i-remove-the-auto-added-supportedruntime-element) that theory.

We don't even ship the config file:

image

**Comment by johnsimons** [comment]
I wonder if we should be setting this property on the csproj - https://docs.microsoft.com/en-us/dotnet/core/tools/csproj#nuspecfile

**Comment by michaelnoonan** [comment]
Yes!

**Comment by michaelnoonan** [comment]
Done.

**Pull request merged**